### PR TITLE
submitting-changes.chapter.md: explain *why* we have three branches

### PR DESCRIPTION
**Pull request by ghost**
The manual does an okay job of explaining the rules for each of the three development branches, but really doesn't give any intuition as to why there are three (why not four? or two?) or how we got where we are today.  This commit tries to fix that.

I have also removed the term "stabilization" from the arc labels. This vague term is not defined anywhere, and does communicate any useful information without a longer explanation.  Therefore it is not appropriate for use in a diagram.

screenshot-20230412-103006

### CODE REVIEW

**Review by teutat3s — APPROVED** [review]
LGTM

File: doc/contributing/submitting-changes.chapter.md [hunk 1]
```diff
@@ -214,47 +214,87 @@ The last checkbox is fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blo
...
+### Branches
+
+The nixpkgs repo has three major branches: `master`, `staging`, and
+`staging-next`.
+
+The most important distinction between them is that `staging`
+(colored red in the diagram below) can receive commits which cause
+a mass-rebuild (for example, anything that changes the `drvPath` of
+`stdenv`).  The other two branches `staging-next` and `master`
+(colored green in the diagram below) can *not* receive commits which
```
Comment (fricklerhandwerk) [inline doc/contributing/submitting-changes.chapter.md:226]: Colors are transient and not even visible to everyone, depending on output medium or anatomy. Also another place for double book keeping to keep the diagram and the text in sync.

We should clarify what a mass rebuild is without going into implementation details that aren't relevant here. Tried to incorporate that into the suggestion.

```suggestion
can receive commits that cause
a mass-rebuild, that is, changes to data a large amount of derivations depend on and therefore require those to be rebuilt.
The other two branches, `staging-next` and `master`, must *not* receive commits which
```

File: doc/contributing/submitting-changes.chapter.md [hunk 2]
```diff
@@ -214,47 +214,87 @@ The last checkbox is fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blo
...
+
+Notice that the automatic merges are all green arrows.  This is by
+design.  Any merge which might cause a mass rebuild on a branch
+which has automatic builds (`staging-next`, `master`) will be a
+manual merge to make sure it is good use of compute power.
+
+Nixpkgs has two branches so that there is one branch (`staging`)
+which accepts mass-rebuilding commits, and one fast-rebuilding
+branch which accepts independent PRs (`master`).  The `staging-next`
+branch allows the Hydra operators to "batch up" groups of commits to
```
Comment (fricklerhandwerk) [inline doc/contributing/submitting-changes.chapter.md:245]:
```suggestion
branch allows the Hydra operators to batch groups of commits to
```

File: doc/contributing/submitting-changes.chapter.md [hunk 3]
```diff
@@ -214,47 +214,87 @@ The last checkbox is fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blo
...
+`staging` to be built.  By keeping the `staging-next` branch
+separate from `staging`, this "batching" decision does not block
```
Comment (fricklerhandwerk) [inline doc/contributing/submitting-changes.chapter.md:247]:
```suggestion
separate from `staging`, this batching does not block
```

Nothing peculiar about batching that seems to warrant putting it in quotes.

**Review by fricklerhandwerk — CHANGES_REQUESTED** [review]
That's a very valuable addition, but I think we should restructure it a bit to make more probable that readers build the right mental model of what's going on, and we can do that by constructing a clearer narrative. Therefore I didn't go into detailed review, and while those suggestions apply in general, they may become invalid in their specific form if the text structure changes. (It would be sensible to incorporate them first and then continuing work on the contents.)

Right now the problem I see is that we're introducing at least three distinct concepts:
- mass rebuilds
- the branch structure
- merge and batch strategies

But the text begins with the branches as if it were the key concept, and otherwise lumps everything together and interleaves subjects without a clear progression. Really what the whole thing is about is an optimization of resource use, and the setup is a particular solution motivated by the phenomenon of mass rebuilds. The text should reflect that and present the information it already contains in the following order:
1. what mass rebuilds are (caused by change in widely used build inputs) and why it's a problem (extreme resource use and delays in fixing broken builds, whatever, you're the expert on this)
2. which solution Nixpkgs maintainers came up with (perform mass rebuilds on separate branch, batch them up on intermediate branch)
3. what these branches are called and how exactly changes progress through that system
4. illustration and diagram description

It probably makes sense to leave the first paragraph as the top level information one will primarily care about to look up the branch names, as a reminder, and then go into the sequence I proposed.

Also please use one line per sentence, it's much easier to make focused suggestions that way. The diffs become unwieldy quickly if lines are broken at arbitrary points.

File: doc/contributing/submitting-changes.chapter.md (hunk 1 above)
Comment (amjoseph-nixpkgs) [reply doc/contributing/submitting-changes.chapter.md:228]: Yes, I'm aware that colorblind people exist, which is one of the reasons why this PR switches from *pastels* to primary colors.

Mentioning the fact that certain parts of the diagram are certain colors does not hurt, even if a few people cannot see those colors.

**Comment by amjoseph-nixpkgs** [comment]
Committed @fricklerhandwerk's suggested changes.

Rebased to fix merge conflict.

No other changes.

**Comment by amjoseph-nixpkgs** [comment]
Squashed.

**Comment by amjoseph-nixpkgs** [comment]
> Really what the whole thing is about is an optimization of resource use, and the setup is a particular solution motivated by the phenomenon of mass rebuilds. The text should reflect that and present the information it already contains in the following order

@fricklerhandwerk, you make several good suggestions on how to improve the prose here.  However a mediocre explanation (this PR) is better than no explanation at all (which is what is currently in-tree).

I've committed all of your suggested changes except the one that deleted the color legend.

The remainder of your comments are suggestions on how to make the prose in this PR better, rather than indicating that the explanation is worse than no explanation at all.  I'm going to merge this PR now so that I don't have to keep rebasing it.  If you would like to submit a follow-up PR to improve the wording I would be happy to review it.

**Comment by amjoseph-nixpkgs** [comment]
I panic-reverted this thinking that I had broken CI on master, but on closer inspection it seems GitHub was putting big red Xes in my notifications list for stale versions of this PR.

Resubmitted as https://github.com/NixOS/nixpkgs/pull/243140

**Comment by lilyinstarlight** [comment]
Did this intentionally revert @pennae's diff from https://github.com/NixOS/nixpkgs/pull/239636/files#diff-ba881223dddc816723e79c2ee074df7b18a08b6520273c22665143b700f3f52dL217-R221 that moved the graphviz dot file out of the markdown?

The graph currently is not rendered on https://nixos.org/manual/nixpkgs/unstable/#submitting-changes-branches as a result of this change

**Comment by amjoseph-nixpkgs** [comment]
> The graph currently is not rendered on https://nixos.org/manual/nixpkgs/unstable/#submitting-changes-branches

Why is inline graphviz not rendered anymore?

**Comment by lilyinstarlight** [comment]
> Why is inline graphviz not rendered anymore?

Presumably because no more docbook, but idk exactly. If you would rather, I suppose I can make the PR that moves this graphviz to the adjacent file and restores @pennae's change

**Comment by infinisil** [comment]
I'm working on moving the contributing section out of the rendered manual and into CONTRIBUTING.md (see https://github.com/NixOS/nixpkgs/issues/244056), where we can then use [GitHub markdowns mermaid support](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/).

So we don't need to spend time trying to fix it in the rendered manual.

**Comment by infinisil** [comment]
```mermaid

```

**Pull request merged**